### PR TITLE
Preparatory refactors: managed expression subject references

### PR DESCRIPTION
**Pull request by samuelhwilliams**
## 🎫 Ticket
https://mhclgdigital.atlassian.net/browse/FSPT-1284

## 📝 Description
This is a backwards-compatible set of refactors to managed expressions that starts moving us away from expecting them to be centred around a `Question` and towards allowing them to reference any piece of data in the expression context.

This introduces a slim `ExpressionReference` type that is essentially just a string. This is used when we are passing around a reference to a datapoint in expression context to help enforce typing through the codebase and provide a single place for working out if a datapoint refers to a question, a data source, or (future) something else. It also has helpers for getting "wrapped" and "unwrapped" references (with/without the `(( ))` we use for interpolation). This should help stop that implementation detail from leaking across the codebase.

Note for the future: I think we should also add something like an ExpressionStatement, which would be used for representing entire statements like `d_123.c_allocation >= q_123 + q_234` and then have helpers for extracting all of the references from it.

After this patch, managed expressions will work with both `question_id`s directly (current state of the world), but also start working with the new `subject_reference` - which can be any context datapoint. Future PRs in this stack will migrate us to `subject_reference` and clean up the old `question_id` limitation.

## 🧪 Testing
This should be a fully-backwards-compatible zero-downtime deploy. After this is merged, any new managed expressions or edited managed expressions should start writing a `subject_reference` into the `context` column of the `expression` table. Future PRs will backfill that key into all rows and then start reading those values instead of `question_id`.

### Review Readiness
- [x] PR title and description provide sufficient context for reviewers
- [x] Commits are logical, self-contained, and have clear descriptions

### Testing
- [x] I have tested this change and it meets the acceptance criteria for the ticket
- [x] I need the reviewer(s) to pull and run this change locally
- [x] New (non-developer) functionality has appropriate unit and integration tests
- [-] End-to-end tests have been updated (if applicable)
- [-] Edge cases and error conditions are tested

### CODE REVIEW

File: app/common/data/interfaces/collections.py [hunk 1]
```diff
@@ -1742,14 +1748,12 @@ def _validate_and_sync_expression_references(expression: Expression) -> None:  #
                 question_to_test=expr_impl.referenced_question if expression.is_managed else None,  # ty:ignore[unresolved-attribute]
             )
 
-            if question_id := SafeQidMixin.safe_qid_to_id(valid_reference):
-                referenced_question = get_question_by_id(question_id)
+            if referenced_question := valid_reference.question:
                 referenced_questions.add(referenced_question)
                 continue
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/common/data/interfaces/collections.py:1753]: SQLAlchemy will cache the question for us - this is ok for now.

File: app/common/expressions/managed.py [hunk 2]
```diff
@@ -487,7 +497,9 @@ def get_form_fields(referenced_question: Question, expression: TOptional[Express
             ),
             "between_bottom_of_range_expression": StringField(
                 "Minimum value",
-                default=expression.context.get("minimum_expression") or "" if expression else "",  # type: ignore[arg-type]
+                default=ExpressionReference(str(minimum_expression)).wrapped
+                if expression and (minimum_expression := expression.context.get("minimum_expression"))
+                else "",
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/common/expressions/managed.py:502]: This will only affect form designers mid-adding context when the deployment happens; we'll live with this for now.

File: app/common/expressions/managed.py [hunk 3]
```diff
@@ -845,7 +851,9 @@ def get_form_fields(referenced_question: Question, expression: TOptional[Express
             ),
             "latest_expression": StringField(
                 "Latest date",
-                default=cast(str, expression.context.get("latest_expression") or "" if expression else ""),
+                default=ExpressionReference(str(latest_expression)).wrapped
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/common/expressions/managed.py:854]: as previous - we'll live with this transient deployment risk

File: app/common/expressions/managed.py [hunk 4]
```diff
@@ -1133,7 +1140,9 @@ def get_form_fields(referenced_question: Question, expression: TOptional[Express
             ),
             "between_top_of_range_expression": StringField(
                 "Latest date",
-                default=expression.context.get("latest_expression") or "" if expression else "",  # type: ignore[arg-type]
+                default=ExpressionReference(str(latest_expression)).wrapped
+                if expression and (latest_expression := expression.context.get("latest_expression"))
+                else "",
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/common/expressions/managed.py:1145]: as previous - we'll live with this transient deployment risk

File: app/common/expressions/references.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            )
+        return super().__new__(cls, value)
+
+    @classmethod
+    def from_wrapped(cls, text: str) -> ExpressionReference:
+        match = _WRAPPED_REFERENCE_RE.match(text)
+        if not match:
+            raise ValueError(f"Expected a wrapped reference like ((ref)); got {text!r}")
+        ref = match.group(1).strip()
+        # TODO: This should raise InvalidReferenceInExpression if an empty string?
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/common/expressions/references.py:54]: as previous - we'll live with this transient deployment risk

File: app/common/expressions/managed.py [hunk 6]
```diff
@@ -266,7 +276,9 @@ def get_form_fields(
             ),
             "greater_than_expression": StringField(
                 "Minimum value",
-                default=expression.context.get("minimum_expression") or "" if expression else "",  # type: ignore[arg-type]
+                default=ExpressionReference(str(minimum_expression)).wrapped
+                if expression and (minimum_expression := expression.context.get("minimum_expression"))
+                else "",
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/common/expressions/managed.py:281]: as previous - we'll live with this transient deployment risk

File: app/common/expressions/managed.py [hunk 7]
```diff
@@ -369,7 +380,9 @@ def get_form_fields(referenced_question: Question, expression: TOptional[Express
             ),
             "less_than_expression": StringField(
                 "Maximum value",
-                default=expression.context.get("maximum_expression") or "" if expression else "",  # type: ignore[arg-type]
+                default=ExpressionReference(str(maximum_expression)).wrapped
+                if expression and (maximum_expression := expression.context.get("maximum_expression"))
+                else "",
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/common/expressions/managed.py:385]: as previous - we'll live with this transient deployment risk

File: app/common/expressions/managed.py [hunk 8]
```diff
@@ -506,7 +518,9 @@ def get_form_fields(referenced_question: Question, expression: TOptional[Express
             ),
             "between_top_of_range_expression": StringField(
                 "Maximum value",
-                default=expression.context.get("maximum_expression") or "" if expression else "",  # type: ignore[arg-type]
+                default=ExpressionReference(str(maximum_expression)).wrapped
+                if expression and (maximum_expression := expression.context.get("maximum_expression"))
+                else "",
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/common/expressions/managed.py:523]: as previous - we'll live with this transient deployment risk

File: app/common/expressions/managed.py [hunk 9]
```diff
@@ -963,7 +969,9 @@ def get_form_fields(referenced_question: Question, expression: TOptional[Express
             ),
             "earliest_expression": StringField(
                 "Earliest date",
-                default=cast(str, expression.context.get("earliest_expression") or "" if expression else ""),
+                default=ExpressionReference(str(earliest_expression)).wrapped
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/common/expressions/managed.py:972]: as previous - we'll live with this transient deployment risk

File: app/common/expressions/managed.py [hunk 10]
```diff
@@ -1112,7 +1117,9 @@ def get_form_fields(referenced_question: Question, expression: TOptional[Express
             ),
             "between_bottom_of_range_expression": StringField(
                 "Earliest date",
-                default=expression.context.get("earliest_expression") or "" if expression else "",  # type: ignore[arg-type]
+                default=ExpressionReference(str(earliest_expression)).wrapped
+                if expression and (earliest_expression := expression.context.get("earliest_expression"))
+                else "",
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/common/expressions/managed.py:1122]: as previous - we'll live with this transient deployment risk

**Review by MarcUsher — APPROVED** [review]
Pair reviewed along with @srh-sloan, tested locally for pre-made managed expressions and all still working as expected 🙌 

Really good test coverage as well and talking through the future PRs all make sense 👌
